src/nupic/algorithms/SpatialPooler.cpp

```diff
@@ -954,53 +955,33 @@ void SpatialPooler::inhibitColumns_(const vector<Real> &overlaps,
   }
 }
 
-bool SpatialPooler::isWinner_(const Real score, const vector<pair<UInt, Real> >& winners,
-                              const UInt numWinners) const {
-  if (score < stimulusThreshold_) { //does not pass threshold, cannot win
-    return false;
-  }
-
-  NTA_ASSERT(numWinners > 0);
-  if (winners.size() < numWinners) { //we haven't populated enough winners yet, accept everybody
-    return true;
-  }
-
-  if (!winners.empty() && score >= winners.back().second) { //winners full, see if better than last accepted, replace
-    return true;
-  }
-
-  return false;
-}
-
-void SpatialPooler::addToWinners_(const UInt index, const Real score,
-                                  vector<pair<UInt, Real>> &winners) const {
-  const pair<UInt, Real> val = make_pair(index, score);
-  for (auto it = winners.begin(); it != winners.end(); it++) {
-    if (score >= it->second) {
-      winners.insert(it, val);
-      return;
-    }
-  }
-  winners.push_back(val);
-}
-
 void SpatialPooler::inhibitColumnsGlobal_(const vector<Real> &overlaps,
                                           Real density,
                                           vector<UInt> &activeColumns) {
   activeColumns.clear();
   const UInt numDesired = (UInt)(density * numColumns_);
   NTA_CHECK(numDesired > 0) << "Not enough columns (" << numColumns_ << ") "
                             << "for desired density (" << density << ").";
-  vector<pair<UInt, Real>> winners;
-  for (UInt i = 0; i < numColumns_; i++) {
-    if (isWinner_(overlaps[i], winners, numDesired)) {
-      addToWinners_(i, overlaps[i], winners);
-    }
-  }
 
-  const UInt numActual = min(numDesired, (UInt)winners.size());
-  for (UInt i = 0; i < numActual; i++) {
-    activeColumns.push_back(winners[i].first);
+  activeColumns.reserve(numColumns_);
+  for(UInt i = 0; i < numColumns_; i++)
+    activeColumns.push_back(i);
+  auto compare = [&overlaps](const UInt &a, const UInt &b) -> bool
+    {return overlaps[a] > overlaps[b];};
+  std::nth_element(
+    activeColumns.begin(),
+    activeColumns.begin() + numDesired,
+    activeColumns.end(),
+    compare);
+  activeColumns.resize(numDesired);
+  std::sort(activeColumns.begin(), activeColumns.end(), compare);
+
+  // Remove sub-threshold winners
+  while(!activeColumns.empty()) {
```

Comment:
I'd use `while(!activeColumns.empty() && overlaps[activeColumns.back()] < stimulusThreshold_)`
and you could avoid the if-else